Use metadata informers in controllers

linxiulei opened the pull request:
Generic informers cache full objects which causes high memory overhead. All generic controllers only use metadata except for resourcequota.

Signed-off-by: Eric Lin <exlin@google.com>

#### What type of PR is this?

/kind bug


#### What this PR does / why we need it:

Improve KCM memory efficiency

#### Which issue(s) this PR fixes:

Fixes #115699

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

aojea commented:
/ok-to-test

pacoxu commented:
/priority important-longterm
The change seems to be valid. LGTM overall.

liggitt commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -234,7 +234,7 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 		}
 
 		controllerContext.InformerFactory.Start(stopCh)
-		controllerContext.ObjectOrMetadataInformerFactory.Start(stopCh)
```
does this mean that if k-c-m was already watching all pods for typed controller loops that required all pod info, this would now *additionally* watch all metadata for pods? in other words, will this double up watching both full info and metadata for some objects?

linxiulei replied:
Yes, but it should not be a big issue in reality because metadata are small so double of metadata is still smaller than having all objects in caches. The problem here -- resourcequota is a generic controller that caches all k8s objects including CRD. It's very likely to cache with excessive memory cost.

liggitt replied:
resourcequota wouldn't watch CRDs, those are cluster-scoped. It also wouldn't watch full CR objects, since we don't have typed informers for those, so all custom resources would already have been falling back to the metadata informer.

If we already have a typed informer running for a given resource type, this change is going to be a net negative for that type. This is only beneficial for resources which we have typed informers for which would not run otherwise. Do we know what resources those are, and whether those objects are typically high volume or large in size?

linxiulei replied:
You are right about CRD and CR objects.

I was thinking ConfigMap/Secret, which are usually larger than their metadata and not watched by other controllers by default.

liggitt replied:
the serviceaccount token controller starts an informer which watches secrets, since it has to populate them with tokens and clean up token secrets for delete serviceaccount objects:

https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/controllermanager.go#L667

the root CA publisher controller starts an informer which watches configmaps, since it has to react to modifications or deletion of the root-ca.crt configmap in every namespace:

https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/certificates.go#L183

linxiulei replied:
Those two controllers are both in my following commits to remove the typed informers for ConfigMap & Secret. But they also need to change other parts of code so I didn't put them in this PR to avoid making it too complicated.

For discussion, we can skip those two controllers and the other two disabled by default  https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/controllermanager.go#L412

liggitt replied:
this change would be less fragile and an unconditional win if it kept the fallback informer factory but adjusted its behavior from "return a typed informer falling back to a metadata informer" to "return an already-started typed informer falling back to a metadata informer"

that would let us use lightweight metadata informers when we didn't already have a running typed informer, and would mean that any follow-up cleanups to eliminate overly broad watching of typed objects would automatically let the generic informers scope down their use to metadata-only informers

I don't think we should drop the fallback and spawn parallel metadata informers for typed things we're already watching

linxiulei replied:
Changed as suggested. PTAL, thanks!

cici37 commented:
/triage accepted

liggitt commented on cmd/kube-controller-manager/app/apps.go:
```diff
@@ -33,6 +35,10 @@ import (
 )
 
 func startDaemonSetController(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {
+	controllerContext.ObjectOrMetadataInformerFactory.WithResource(appsv1.SchemeGroupVersion.WithResource("daemonsets"))
+	controllerContext.ObjectOrMetadataInformerFactory.WithResource(appsv1.SchemeGroupVersion.WithResource("controllerrevisions"))
+	controllerContext.ObjectOrMetadataInformerFactory.WithResource(corev1.SchemeGroupVersion.WithResource("pods"))
+	controllerContext.ObjectOrMetadataInformerFactory.WithResource(corev1.SchemeGroupVersion.WithResource("nodes"))
```
hmm... this approach is really fragile to drift/typos/omission... I was expecting to either augment / wrap `controllerContext.InformerFactory` to track what typed informers were requested, or adjust the generated informerfactories to track group/version/resource of requested informers, and propagate that info to the started-typed-or-metadata factory

liggitt commented on pkg/quota/v1/evaluator/core/registry.go:
```diff
@@ -32,6 +32,15 @@ var legacyObjectCountAliases = map[schema.GroupVersionResource]corev1.ResourceNa
 	corev1.SchemeGroupVersion.WithResource("secrets"):                corev1.ResourceSecrets,
 }
 
+func NeedFullObjectForResource(resource schema.GroupVersionResource) bool {
+	switch resource.GroupResource() {
+	case schema.GroupResource{Resource: "pods"}, schema.GroupResource{Resource: "services"}, schema.GroupResource{Resource: "persistentvolumeclaims"}:
+		return true
+	default:
+		return false
+	}
+}
+
```
can this be included in the `quota.Evaluator` interface so podEvaluator / serviceEvaluator / persistentVolumeClaim evaluator answer whether it requires the full object or not, and we don't ever drift?

linxiulei replied:
The evaluator might be created after this function (see pkg/controller/resourcequota/resource_quota_monitor.go:controllerFor() ), it seems infeasible to me. At least not without major changes.

linxiulei commented:
/retest

liggitt committed:
Track group/version/resource for requested informers

linxiulei committed:
Run hack/update-codegen.sh informers subprojects

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei committed:
Use metadata informers in controllers

This commit makes controllers use metadata informers if corresponding
typed informers are not created to decrease memory usage.

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -414,6 +414,13 @@ var ControllersDisabledByDefault = sets.NewString(
+// GenericControllers is the set of controllers which watch all resources. Generic controllers are started
+// after others to reuse typed informers initialized by others.
+var GenericControllers = sets.NewString(
+	"resourcequota",
```
I think we can use WaitGroup here.

For controllers except for GC and ResourceQuota:
```
wg.Add(len(controllersToStart))

# for each controller
StartController()
wg.Done()
```

For GC and ResourceQuota:

```
# pass wg into controller initialization func
wg.Add(1)
startResourceQuotaController(wg)

# inside startResourceQuotaController()
wg.Wait()
startDiscoveryAndCreateMetadataInformers()
```

But to use WG, we have to make initialization functions of GC and ResourceQuota non-blocking. WDYT?

linxiulei replied:
Now that the code has changed a bit and that makes the starting order easier to manipulate. However,  it's still not like what you suggested because of the complexity. Anyway, please have a look.

linxiulei commented on pkg/controller/resourcequota/resource_quota_monitor.go:
```diff
@@ -164,7 +166,15 @@ func (qm *QuotaMonitor) controllerFor(ctx context.Context, resource schema.Group
+		shared informers.GenericInformer
+		err    error
+	)
+	if core.NeedFullObjectForResource(resource) {
```
Sounds great. Thanks!

linxiulei commented:
/retest

linxiulei commented:
/remove-lifecycle stale

linxiulei commented:
/retest
